debian/man/faxmsg.8
.TH FAXMSG 8
.SH NAME
faxmsg \- back end to faxabort, faxanswer and faxquit
.SH SYNOPSIS
.B /usr/sbin/faxmsg
.SH "DESCRIPTION"
.B faxmsg
is never called directly - it is always called by the names
.B faxabort, faxanswer,
or
.B faxquit.

This manual page was written for the Debian GNU/Linux distribution
(but may be used by others), because the original program does not
have a manual page.
.SH "SEE ALSO"
.IR hylafax (5),
.IR faxabort (8),
.IR faxanswer (8),
.IR faxquit (8)
.SH AUTHOR
This manual page was written by Dermot Bradley <bradley@debian.org>,
for the Debian GNU/Linux system.


CHANGES
#	$Id$	

Chandlog since HylaFAX 4.3.1

* PAM reworked to use complete sessions independantly (25 Jan 2007)
* Added debian/ directory with rules to create a simple Debian package (21 Jan 2007)
* Simplify/Unify mime encoder selection and use (9 Jan 2007)
* Add the Include: option for Config file options (8 Jan 2007)
* Redirected unwanted output from pd2pdf while running notify (24 Dec 2006)
* Use @SCRIPT_SH@ for bin/ scripts to allow configure selection (23 Dec 2006)
* Updated pollrcvd to use templates (22 Dec 2006)
* Moved default templates to etc/tempates/en/ to clean up etc/templates/ (22 Dec 2006)

Changelog for HylaFAX 4.3.1

* Added exit codes to faxalter (Bug 796) (4 Dec 2006)
* Fixed faxqclean to work correctly with cover documents using the new naming
  scheme, yet still work with previously named documents (bug 831) (20 Nov 2006)
* Fixed sendfax PDF page counting (bug 830) (13 Nov 2006)
* Disable JPEG support when compiling with older libtiff (3 Nov 2006)
* Use old style lock files for OpenServer 6 (20 Oct 2006)
* Always update on-disk qfiles for jobs when state changes to running (19 Oct 2006)
* Change how we compile shared objects for Solaris DSO support (12 Oct 2006)
* Base64 encoding with uuencode was broken on 3 Oct (6 Oct 2006)
* uuencode STDERR was not redirected to $ERRORSTO (6 Oct 2006)
* There was a small possibility of corruption a base64 output produced
  by uuencode if an encoded line started with "begin" (6 Oct 2006)
* Private tmpdir was not set by faxrcvd, but used for body encoding (6 Oct 2006)
* fax faxqclean, don't remove cover documents still referenced (4 Oct 2006)
* Encodes email body according to template setting (3 Oct 2006)
* Add Quoted-Printable encoding option (3 Oct 2006)
* Add Romanian templates (29 Sep 2006)
* faxsetup nolonger blindly removes $TMPDIR (29 Sep 2006)
* Add Motorola SM56 USB (Apple) modem (27 Sep 2006)
* Add HPUX 11.11 support to configure (27 Sep 2006)
* make cover pages visible in /docq/ through hfaxd (17 Sep 2006)
* handle +FHNG/+FNG after AT+FDT in Class 2 (15 Sep 2006)
* Unify mime encoding between notify and faxrcvd
* Add templates for use in faxrcvd translations (14 Aug 2006)
* don't exit faxq with assertion failure when a job times out and is
  interrupted during sending (21 Jul 2006)
* faxaddmodem: Allow identification when no AT+FMM? or AT+FMI? (21 Jul 2006)
* fix error of accepting too-short of DCS signal (20 Jul 2006)
* fix faxgetty hang in ECM due to missing timeouts (20 Jul 2006)
* fix indicator when 100% non-zero TCF data is received (20 Jul 2006)
* do Class1SwitchingCmd before sending MCF after PPS-NULL (20 Jul 2006)
* replace Class1SendMsgDelay with Class1SwitchingCmd (20 Jul 2006)
* fix faxaddmodem config parsing for Digicom Leonardo modems(19 Jul 2006)
* don't skip initial NSF bytes in station ID search when no
  known T.35 manufacturer ID was used (18 Jul 2006)
* Documentation updates (18 Jul 2006)
* Fix RecvFmt string printing (Bug 805) (18 Jul 2006)
* fix timeout on AT command sending to modem (5 Jul 2006)
* don't use tiffcp "-r" option on JBIG TIFFs (5 Jan 2006)
* Avoid problems when resending single ECM Frame (5 Jul 2005)
* Improve recovery after Class 1 TCF receipt times out (4 Jul 2006)
* Improve the handling of user aborts in Class 1 sending (4 Jul 2006)
* Add missing From: headers to notify (4 Jul 2006)
* correct CIELAB to be ITULAB (4 Jul 2006)
* fix Class 2 data format parsing of +FIS response (4 Jul 2006)
* Enhance faxinfo with formating options (2 Jul 2006)
* Fix faxmail's handling of nested mime (14 Jun 2006)
* Add JobProtection option to hfaxd (12 Jun 2006)
* Enhance SHIELDED_DTMF (5 Jun 2006)
* Add resubmit capability to faxalter (2 Jun 2006)
* Use ] && [ instead of -a for shell portability (2 Jun 2006)
* Add continuation support to FaxClient in libfaxutil (31 May 2006)
* Add UseJobTagLine config option for faxsend  (4 Apr 2006)

Changelog for HylaFAX 4.3.0

* Fix DCS tag handling in Class2 (BUG 771) (5 May 2006)
* Fix hfaxd corruption on killed jobs (BUG 401) (5 May 2006)
* More fixes to batching and jobcontrol interaction (3 May 2006)
* Re-work fix for hfaxd's time handling (BUG 723) (28 Apr 2006)
* Don't let users kill each other's jobs (BUG 752) (28 Apr 2006)
* Use areBatchable() to make sure batching doesn't ignore job
  modems (28 Apr 2006)
* Enhance Italian translation (BUG 758) (26 Apr 2006)
* Fix faxq sleepq concurrency problem (BUG 745) (26 Apr 2006)
* Fix runSchedule() concurrency problem (BUG 745) (26 Apr 2006)
* Fix some of the error paths in the setReadyToRun jobcontrol 
  pipe, fork, and exec calls (BUG 745) (25 Apr 2006)
* Fix unblocking job problem (BUG 733) (24 Apr 2006)
* back out flush modem I/O (BUG 756) (24 Apr 2006)
* Make sure hfaxd returns JBIG faxes in recvq list (BUG 731) (23 Apr 2006)
* Resolve Solaris bus error under native compiler (BUG 769) (23 Apr 2006)
* Fix quoted-printable MIME encoding in faxrcvd (BUG 760) (21 Apr 2006)
* Fix sendfax to accept both destination and user subaddresses
  (BUG 762) (21 Apr 2006)
* Repair broken MMR from some Multitech modems (BUG 763) (21 Apr 2006)
* Add JobControlWait to allow faxq to wait for JobControlCmd
  to finish synchronously (4 Apr 2006)
* flush modem I/O on receptions before sending any data (03 Apr 2006)
* fix incorrect "Fax protocol error" in ECM mode (03 Apr 2006)
* add German NSFs and correct HylaFAX NSF bit order (03 Apr 2006)
* set Class1RMPersistence to 0 in digi config prototype (03 Apr 2006)
* extend V.21 HDLC frame reception timeout to 10 sec (03 Apr 2006)
* ignore MESSAGE-WAITING response after dialing (02 Apr 2006)
* Correctly remove EOFB at the end of MMR images (02,10 Apr 2006)
* set minimum of 4800 for V.34 primary rate renegotiations (02 Apr 2006)
* improve handling of V.34 control channel retrain (02 Apr 2006)
* Add isdn4linux modem config prototype (02 Apr 2006)
* Add IAXmodem config prototype (02 Apr 2006)
* Update config.guess & config.sub (27 Mar 2006)
* Add JobControl (obsoletes DestControl) (27 Mar 2006)

util/faxmsg.c
/*	$Id$ */
/*
 * Copyright (c) 1990-1996 Sam Leffler
 * Copyright (c) 1991-1996 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */
#include <stdio.h>
#include <string.h>
#include <syslog.h>
#include <stdarg.h>
#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>

#include "config.h"
#include "port.h"

static void
fatal(char* fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    if (isatty(fileno(stderr))) {
        vfprintf(stderr, fmt, ap);
        putc('\n', stderr);
    } else {
        vsyslog(LOG_ERR, fmt, ap);
    }
    va_end(ap);
    exit(-1);
}

extern	int cvtFacility(const char*, int*);

int
main(int argc, char** argv)
{
    extern int optind;
    extern char* optarg;
    int fifo, c;
    char* spooldir = FAX_SPOOLDIR;
    const char* arg = NULL;
    char fifoname[256];
    int cmdlen, fnlen;
    char cmd[80];
    char* appname;
    const char* opts = NULL;
    const char* usage = NULL;
    const char* cmdfmt = NULL;
    char* cp;
    int facility = LOG_DAEMON;
    int modemRequired = 0;

    (void) cvtFacility(LOG_FAX, &facility);
    openlog(argv[0], LOG_PID|LOG_ODELAY, facility);
    appname = strrchr(argv[0], '/');
    if (appname)
	appname++;
    else
	appname = argv[0];
    if (strcmp(appname, "faxanswer") == 0) {
	opts = "h:q:";
	usage = "[-h how] [-q queue-dir] modem";
	cmdfmt = "A%s";
	arg = "";			/* default how */
	modemRequired = 1;
    } else if (strcmp(appname, "faxquit") == 0) {
	opts = "q:";
	usage = "[-q queue-dir] [modem]";
	cmdfmt = "Q";
	modemRequired = 0;
    } else if (strcmp(appname, "faxabort") == 0) {
	opts = "q:";
	usage = "[-q queue-dir] modem";
	cmdfmt = "Z";
	modemRequired = 1;
    } else if (strcmp(appname, "faxlock") == 0) {
	opts = "q:";
	usage = "[-q queue-dir] modem";
	cmdfmt = "L";
	modemRequired = 1;
    } else {
	fatal("Unrecognized command name %s", appname);
    }
    while ((c = getopt(argc, argv, opts)) != -1)
	switch (c) {
	case 'h':
	    arg = optarg;
	    break;
	case 'q':
	    spooldir = optarg;
	    break;
	case '?':
	    fatal("usage: %s %s", argv[0], usage);
	    /*NOTREACHED*/
	}
    if (optind == argc-1) {
        if (argv[optind][0] == FAX_FIFO[0]) {
            if (strlen(argv[optind]) < sizeof(fifoname)) {
                strcpy(fifoname, argv[optind]);
            } else {
                fatal("Argument name too long: %s", argv[optind]);
            }
        } else {
            fnlen = snprintf(fifoname, sizeof(fifoname), "%s.%.*s", FAX_FIFO,
                (int) (sizeof(fifoname) - sizeof(FAX_FIFO)), argv[optind]);
            if (fnlen < 0 || fnlen >= sizeof(fifoname)) {
                fatal("Argument name too long: %s", argv[optind]);
            }
        }
    } else if (!modemRequired) {
        strcpy(fifoname, FAX_FIFO);
    } else {
        fatal("usage: %s %s", argv[0], usage);
    }
    for (cp = fifoname; (cp = strchr(cp, '/')); *cp++ = '_')
	;
    if (chdir(spooldir) < 0) {
        fatal("%s: chdir: %s", spooldir, strerror(errno));
    }
    fifo = open(fifoname, O_WRONLY|O_NDELAY);
    if (fifo < 0) {
        fatal("%s: open: %s", fifoname, strerror(errno));
    }
    cmdlen = snprintf(cmd, sizeof(cmd), cmdfmt, arg);
    if (write(fifo, cmd, cmdlen) != cmdlen) {
        fatal("FIFO write failed for command (%s)", strerror(errno));
    }
    (void) close(fifo);
    return 0;
}
